### PR TITLE
Additions and enhancements to the JavaScript API

### PR DESCRIPTION
**Pull request by victor-dumitrescu**
This PR contains:

* Bindings for more existing wasm functions in api.json (NaCl, HMAC, keccak)
* Various bug fixes and enhancements to make these bindings work
* Bindings for P-256: relies on merging KreMLin support for extracting P-256 to wasm ([!198](https://github.com/FStarLang/kremlin/pull/198))
* P-256 point compression/decompression tests have also been added to the OCaml API, since they were missing

### CODE REVIEW

**Comment by msprotz** [comment]
Thanks for the PR. Can you explain:
- the new size_offset parameter?
- why the ML tests for P256 have changed?

Thanks!

**Comment by victor-dumitrescu** [comment]
The `size_offset` parameter is an integer that can be added to the `size` parameter to specify the size of a buffer. I needed this in order to specify the binding of the "easy" versions of Box functions, where the tag and the ciphertext are in one buffer, which has to be the size of the plaintext buffer + 16, for example here:
```json
        {
          "name": "c",
          "kind": "output",
          "type": "buffer",
          "size": "mlen",
          "size_offset": 16,
          "tests": [ ]
        },
 ```
I've also documented it [here](https://github.com/project-everest/hacl-star/blob/71fda46c5406a6b47f5adf6036b7a2999c405c1c/bindings/js/api.js#L133). Let me know if you think we should do this some other way.

The P-256 test is the same, I just changed some strings from decimal to hex, in order to use them in api.json. I had originally copied it from the C tests where it was written in a mixture of both and didn't bother changing it at the time.

**Comment by msprotz** [comment]
Ah it's just the name that confused me: I somehow thought that `offset` referred to an offset in memory and implied somehow that the size should be understood after pointer addition of 16 or something.

What do you think of doing something real quick, where size may be either `foo` (a parameter) or `foo+16` (parameter + extra constant size)? You could parse this in a really degenerate way (e.g. `if string.indexOf('+' >= 0) ...`) and it would enhance clarity of the API.

Thanks for the explanation regarding the tests, looks good!

**Comment by victor-dumitrescu** [comment]
I thought of doing that at first but decided against it because of the messier implementation, but maybe it's worth prioritising the clarity of the bindings.

**Comment by msprotz** [comment]
maybe something like this:
```
function parseLength(l) {
  var i = l.indexOf('+');
  if (i < 0)
    return [l,0];
  var base = l.substring(0, i).trim();
  var length = parseInt(l.substring(i+1));
  return [base,length];
}

var [ buf, extra ] = parseLength(...);
```

I don't think the parsing cost is too onerous, what do you think?

**Comment by victor-dumitrescu** [comment]
Just pushed an attempt at this. I actually also need (-) so it turned out a bit more convoluted. Let me know what you think.

File: bindings/js/api.js
Comment (msprotz) [inline bindings/js/api.js:135]: need to update here

File: bindings/js/api.js
Comment (msprotz) [inline bindings/js/api.js:172]: rename to `evalSizeWithOp`

File: bindings/js/api.js
Comment (msprotz) [inline bindings/js/api.js:310]: can we have a more helpful error message?

File: bindings/js/api_test.js
Comment (msprotz) [inline bindings/js/api_test.js:57]: @denismerigoux can you check this?

File: bindings/js/api.js
Comment (msprotz) [inline bindings/js/api.js:180]: typo

**Review by msprotz — APPROVED** [review]
Looks good to me modulo:
- my small remarks
- Denis' approval for the changes to the JS API test harness

Let's wait for @denismerigoux to review this and then we'll be good to go.

Thanks @victor-dumitrescu for improving those bindings!

File: bindings/js/api_test.js
Comment (denismerigoux) [reply bindings/js/api_test.js:57]: So here @victor-dumitrescu you removed a small optimization I had made. Indeed, normally all API functions return an array containing the tuple of the results. But in the case where the tuple had one element, I was removing the array and directly returning the element. Do you remember why you changed this ?

**Review by denismerigoux — APPROVED** [review]
Modulo my small comment, I also approve the changes, thanks Victor!

File: bindings/js/api_test.js
Comment (victor-dumitrescu) [reply bindings/js/api_test.js:57]: There were a few things happening here. All functions in this category return a tuple of (result, output buffers) but we only care about the result at this stage, so it's fine to discard the rest. Here this was done if the function has any output buffers specified in api.json. However, even if that's not the case the function will return a tuple of (result, []) so I eliminated this check and I'm always taking the first part of the tuple as the result. Otherwise in this case result_val would be something like `[ true, [] ]`

Then, the result_val / expected_result check was broken, it should have been an inequality since if it's true we throw an error. The reason this wasn't being triggered is that this was comparing strings (as returned by `postprocessing`) with bool/ints (as returned by the function), which would always be false with === and so the test would pass. I added the posprocessing for the return value as well and fixed the check.

File: bindings/js/api.js
Comment (victor-dumitrescu) [reply bindings/js/api.js:310]: Sorry, I actually didn't mean to introduce this change, it was a test from another version of the file. Thanks for spotting it, I will reverse it.

File: bindings/js/api_test.js
Comment (denismerigoux) [reply bindings/js/api_test.js:57]: Oh thank you, I had not spotted this. Damn Javascript operators... I approve then :)

**Comment by victor-dumitrescu** [comment]
Thanks for the review, I'll add the remaining tests and bindings and then merge.

**Comment by msprotz** [comment]
Thanks Victor! Let's make sure we address Denis' remark here: https://github.com/project-everest/hacl-star/pull/378#discussion_r512521281 before merging. And then we're good.

**Comment by victor-dumitrescu** [comment]
We already addressed it in the inline comment, I think there's no further change required

**Comment by msprotz** [comment]
Ah thanks for some reason the github interface showed two different comment threads for the same issue. All good.

**Comment by denismerigoux** [comment]
Correct, all good to go !

File: bindings/js/api.js
Comment (victor-dumitrescu) [inline bindings/js/api.js:244]: While adding tests for NaCl I had to make one last change to api.json on which I'd like to get @denismerigoux's approval before merging. Here we were flattening the return_buffers array only if it contained exactly one buffer, which is the case for almost all the functions in the API. A function would thus return
1. `[ ret ]` if it has no output buffers
2. `output_buf` if its return type is void and it has 1 output buffer
3. `[ ret, output_buf ]` if its return type is not void and it has 1 output buffer
4. `[ ret, [ output_buf1, output_buf2, ...]]` if its return type is not void and it has more than 1 output buffer

With these changes, case 1 and 3 remain unchanged and case 2 and 4 are now:
2. `[ output_buf ]` if its return type is void and it has 1 output buffer
4. `[ ret, output_buf1, output_buf2, ...]` if its return type is not void and it has more than 1 output buffer

So now, irrespective of return type and number of output values, all JS functions will return an array in which the first element is the return value (absent if return type if void) and the rest of the elements are output buffers.

Comment (victor-dumitrescu) [reply bindings/js/api.js:244]: I've made this change because I think it's a bit cleaner and it simplifies the logic required to consume these functions.

Comment (denismerigoux) [reply bindings/js/api.js:244]: I'm fine with that, it's indeed more regular

**Pull request merged**
